### PR TITLE
feat(schema-migrate): reporter and error handling

### PR DESCRIPTION
**Pull request by Bento007**
## Reason for Change

- #5100

## Changes
- Update the schema-migration step function to handle error cases gracefully
- Store report information under the step function execution ID in s3.
- Clean up S3 report and migration files after running migration.
- Add REMOTE_DEV_PREFIX to S3 files path for rdev compatibility
- Log REMOTE_DEV_PREFIX during dataset processing. 
- Fix `cleanup_artifacts` in upload_failure app to correctly cleanup artifact if processing fails by passing in the correct parameters.
- Schema_migration
- Add logging to schema_migration to simplify debugging
- Add logging for requests to third party services.
- Correct documentation for gather_collections
- Simplify the loop in gather_collections
- Corrected `collection_id` to `collection_version_id` where needed.
- Change variables names in schema_migration.py to be more descriptive.
- Modified `collection_migrate `for better error handling in the step function.
- Call `ingest_dataset` and skipping the processing step function trigger. This sets up the database to process the dataset in the next step of the schema_migration step function.
- Move `can_publish` logic to the top of `collection_migrate`
- Changed `CAN_OPEN_REVISON` to `CAN_PUBLISH`'
- Added `no_datasets` to skip processing a collections when no datasets are present.
- Add combing of reported status
- Add slack handler to upload files to report channel
- Add slack token and slack channel to rdev config secret
- add reporting to slack channel
- related infra changes need to test on rdev https://github.com/chanzuckerberg/single-cell-infra/pull/819

## Testing steps

- [step function run](https://us-west-2.console.aws.amazon.com/states/home?region=us-west-2#/v2/executions/details/arn:aws:states:us-west-2:699936264352:execution:dp-rdev-schema-mig-schema-migration-sfn:0176b7d9-2eea-48fa-bba3-fdcb33278f80)
- Verified the slack out in curation slack reporting channel.

## Notes for Reviewer
- The  prod config secret will need to be updated with the slack channel and slack token.
- need to give rdev access to read from the dev artifact bucket. https://github.com/chanzuckerberg/single-cell-infra/pull/819

### CODE REVIEW

**Review by nayib-jose-gloria — COMMENTED** [review]
one note re: the error decorator, otherwise this looks good and ready to test!

File: backend/layers/business/business.py [hunk 1]
```diff
@@ -402,7 +413,8 @@ def ingest_dataset(
         )
 
         # Starts the step function process
-        self.step_function_provider.start_step_function(collection_version_id, new_dataset_version.version_id, url)
+        if start_step_function:
```
Comment (Bento007) [inline backend/layers/business/business.py:416]: This is so we can run ingest_dataset without starting the submissions pipeline

File: backend/layers/processing/upload_failures/app.py [hunk 2]
```diff
@@ -27,7 +27,7 @@ def handle_failure(event: dict, context) -> None:
         dataset_id, collection_version_id, error_step_name, error_job_id, error_aws_regions, execution_arn
     )
     update_dataset_processing_status_to_failed(dataset_id)
-    cleanup_artifacts(event)
+    cleanup_artifacts(dataset_id, error_step_name)
```
Comment (Bento007) [inline backend/layers/processing/upload_failures/app.py:30]: This is an unrelated fix to the submission error handle.

File: backend/layers/processing/process.py [hunk 3]
```diff
@@ -80,6 +80,7 @@ def log_batch_environment(self):
             "DEPLOYMENT_STAGE",
             "MAX_ATTEMPTS",
             "MIGRATE",
+            "REMOTE_DEV_PREFIX",
```
Comment (Bento007) [inline backend/layers/processing/process.py:83]: for logging

File: backend/layers/business/business.py [hunk 4]
```diff
@@ -329,11 +331,20 @@ def ingest_dataset(
         url: str,
         file_size: Optional[int],
         existing_dataset_version_id: Optional[DatasetVersionId],
+        start_step_function: bool = True,
     ) -> Tuple[DatasetVersionId, DatasetId]:
         """
         Creates a canonical dataset and starts its ingestion by invoking the step function
         If `size` is not provided, it will be inferred automatically
         """
+        logger.info(
```
Comment (Bento007) [inline backend/layers/business/business.py:340]: for logging

File: backend/layers/processing/schema_migration.py [hunk 5]
```diff
@@ -52,127 +56,134 @@ def gather_collections(self) -> List[Dict[str, str]]:
...
+            "dataset_version_id": new_dataset_version_id.id,
+            "uri": uri,
+        }
 
     def collection_migrate(
-        self, collection_id: str, collection_version_id: str, can_open_revision: bool
-    ) -> List[Dict[str, str]]:
-        private_collection_id = collection_version_id
+        self, collection_id: str, collection_version_id: str, can_publish: bool
+    ) -> Union[Dict[str, Any]]:
```
Comment (nayib-jose-gloria) [inline backend/layers/processing/schema_migration.py:106]: nit: the return type doesn't need to be a Union

File: backend/layers/processing/schema_migration.py [hunk 6]
```diff
@@ -18,27 +19,30 @@
...
+from backend.layers.processing.process_logic import ProcessingLogic
 from backend.layers.thirdparty.step_function_provider import StepFunctionProvider
 
 logger.configure_logging(level=logging.INFO)
 
 
-class SchemaMigrate:
+class SchemaMigrate(ProcessingLogic):
     def __init__(self, business_logic: BusinessLogic):
+        # TODO: remove dependecies on ProcessingLogic
```
Comment (nayib-jose-gloria) [inline backend/layers/processing/schema_migration.py:30]: is this TODO for the future?

Comment (Bento007) [reply backend/layers/processing/schema_migration.py:30]: yes, in the event we want to completely decouple from ProcessnigLogic.

File: backend/layers/processing/requirements.txt [hunk 7]
```diff
@@ -1,4 +1,6 @@
-cellxgene-schema
+git+https://github.com/chanzuckerberg/single-cell-curation/@auto/update-convert-py-to-3.1.1#subdirectory=cellxgene_schema_cli
```
[the remark this answers is not in the document]
Comment (Bento007) [reply backend/layers/processing/requirements.txt:1]: It will be removed before moving to dev.

File: backend/layers/processing/schema_migration.py [hunk 8]
```diff
@@ -49,156 +56,205 @@ def gather_collections(self) -> List[Dict[str, str]]:
...
+        migrated_file = "migrated.h5ad"
+        migrate("previous_schema.h5ad", migrated_file, collection_version_id, dataset_id)
+        key_prefix = self.get_key_prefix(dataset_version_id)
+        uri = self.upload_artifact(migrated_file, key_prefix, self.artifact_bucket)
+        new_dataset_version_id, _ = self.business_logic.ingest_dataset(
+            CollectionVersionId(collection_version_id),
+            uri,
+            file_size=0,  # TODO: this shouldn't be needed but it gets around a 404 for HeadObject
+            existing_dataset_version_id=DatasetVersionId(dataset_version_id),
+            start_step_function=False,
```
Comment (nayib-jose-gloria) [inline backend/layers/processing/schema_migration.py:96]: Shouldn't this be True?

Comment (Bento007) [reply backend/layers/processing/schema_migration.py:95]: It should be false. I added a comment.

**Review by nayib-jose-gloria — APPROVED** [review]
